### PR TITLE
[CELEBORN-1742] Fix bounded driver Inbox deadlock

### PR DESCRIPTION
**Pull request by CodingCat**
### What changes were proposed in this pull request?

fix the deadlock by clear the message queue if it has been full before we put OnStop message

### Why are the changes needed?

it is possible that the message queue has been full when we want to insert OnStop, and unfortunately, we have to hold the inbox as the lock when doing this....

this creates the scenario of deadlock, main thread is waiting for more capacity holding inbox lock, message processors are stuck because it needs inbox lock

(the screenshots are based on 0.4 client tho)

main thread:

image

processor thread:

image

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

in prod

### CODE REVIEW

**Review by RexXiong — APPROVED** [review]

**Review by FMX — APPROVED** [review]
LGTM.

**Comment by SteNicholas** [comment]
@CodingCat, please fix the failed test case `RpcEnvSuite`.

**Comment by cfmcgrady** [comment]
Was this bug introduced by https://github.com/apache/celeborn/pull/2366?

**Comment by CodingCat** [comment]
> Was this bug introduced by https://github.com/apache/celeborn/pull/2366?

yes

File: common/src/main/scala/org/apache/celeborn/common/rpc/netty/Inbox.scala [hunk 1]
```diff
@@ -278,6 +278,11 @@ private[celeborn] class Inbox(
         // We should disable concurrent here. Then when RpcEndpoint.onStop is called, it's the only
         // thread that is processing messages. So `RpcEndpoint.onStop` can release its resources
         // safely.
+        // we need to have this check in case that inbox has been full and we have no way to
+        // put OnStop leading the deadlock when shutting down
+        if (capacity > 0 && messageCount.get() >= capacity) {
```
Comment (SteNicholas) [inline common/src/main/scala/org/apache/celeborn/common/rpc/netty/Inbox.scala:283]: Why does this add this check? IMO, we could directly clear the messages.

**Comment by RexXiong** [comment]
It seems current main/branch-0.5 doesn't have this bug, because the messages object of Inbox is using a LinkedList.  But from your screenshot, it seems to be a blocking queue @CodingCat

**Review by turboFei — APPROVED** [review]

File: common/src/main/scala/org/apache/celeborn/common/rpc/netty/Inbox.scala (hunk 1 above)
Comment (CodingCat) [reply common/src/main/scala/org/apache/celeborn/common/rpc/netty/Inbox.scala:280]: it is to ensure that we only clear the messages when the queue is full, additionally, it is to pass ut

**Comment by CodingCat** [comment]
> It seems current main/branch-0.5 doesn't have this bug, because the messages object of Inbox is using a LinkedList. But from your screenshot, it seems to be a blocking queue @CodingCat

actually, I found that it is true that current master does not have this issue, however, it was lucky to have a [bug](https://github.com/apache/celeborn/pull/2943/files#diff-d7586a8a84ac32a463f30dffc72852b6fb8a8722d184ec4872e36e519af7280dR289) which misses the check of the queue length when adding onStop

so I am not entirely sure what the ideal way is to move forward here....

File: common/src/main/scala/org/apache/celeborn/common/rpc/netty/Inbox.scala (hunk 1 above)
Comment (SteNicholas) [reply common/src/main/scala/org/apache/celeborn/common/rpc/netty/Inbox.scala:280]: @CodingCat, the source of truth should be clear the messages regardless of whether the queue is full or not for `stop`.

Comment (RexXiong) [reply common/src/main/scala/org/apache/celeborn/common/rpc/netty/Inbox.scala:280]: > @CodingCat, the source of truth should be that clearing the messages regardless of whether the queue is full or not for `stop`. We do not need to add this check for passing ut.

The current code doesn't seem to have any obvious issues, and I prefer to keep the status, so I will close this PR unless new problems arise. cc @SteNicholas @CodingCat

**Comment by RexXiong** [comment]
The current code doesn't seem to have any obvious issues, and I prefer to keep the status, so I close this PR unless new problems arise.

**Pull request closed without merging**